src/core/thread/network_data_service.cpp

```diff
@@ -57,26 +57,30 @@ Error Manager::AddService(const void *aServiceData,
                           uint8_t     aServiceDataLength,
                           bool        aServerStable,
                           const void *aServerData,
-                          uint8_t     aServerDataLength)
+                          uint8_t     aServerDataLength,
+                          bool        aRegisterWithLeader)
 {
     Error error;
 
     SuccessOrExit(error = Get<Local>().AddService(
                       kThreadEnterpriseNumber, reinterpret_cast<const uint8_t *>(aServiceData), aServiceDataLength,
                       aServerStable, reinterpret_cast<const uint8_t *>(aServerData), aServerDataLength));
 
+    VerifyOrExit(aRegisterWithLeader);
     Get<Notifier>().HandleServerDataUpdated();
```

Comment:
Other network data attributes such as on-mesh prefixes and routes do not call `HandleServerDataUpdate()` and require the user to call it separately. I'm wondering if we should remove the call here to be consistent with other network data methods.

Reply:
I agree. I was initially thinking that this may make it easier for `NetworkData::Service` by taking care of registering with leader from same method, but with `Publisher` taking care of this, it may be easier to go back to the model of caller needing to take care of `HandleServerDataUpdate()` call.

Reply:
Updated in new pushed commit.